### PR TITLE
Migrate StrategyExecutor to new transformer infrastructure

### PR DESCRIPTION
**Pull request by ammareltigani**
Migrated the StrategyExecutor class to strategy_executor() function that uses transformer_primitives. Depreciated StrategyExecutor.

### CODE REVIEW

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 1]
```diff
@@ -100,6 +106,47 @@ def optimization_at(
...
+        raise ArgumentError(execution_strategy, 'Execution strategy cannot be None')
+
+    mapping = execution_strategy.initial_mapping.copy()
+
+    def map_func(op: 'cirq.Operation', index) -> 'cirq.OP_TREE':
+        if isinstance(op.gate, AcquaintanceOpportunityGate):
+            logical_indices = tuple(mapping[q] for q in op.qubits)
+            logical_operations = execution_strategy.get_operations(logical_indices, op.qubits)
+            clear_span = int(not execution_strategy.keep_acquaintance)
+            new_operation = logical_operations.on(*logical_indices)
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:128]: `execution_strategy.get_operations` returns a `cirq.OP_TREE` and therefore `logical_operations` is `cirq.OP_TREE`; which cannot directly be applied on `logical_indices`. 

I think we should directly return `logical_operations` instead of `new_operation` here?

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 2]
```diff
@@ -100,6 +106,47 @@ def optimization_at(
...
 
 
+@cirq.transformer
+def strategy_executor(
+    circuit: 'cirq.Circuit',
+    *,
+    execution_strategy: ExecutionStrategy = None,
+    context: Optional['cirq.TransformerContext'] = None,
+) -> None:
+    """Executes an acquaintance strategy"""
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:116]: Extend the docstring to document the args (`circuit`, `context`, `execution_strategy`) and types of errors it raises.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 3]
```diff
@@ -100,6 +106,47 @@ def optimization_at(
...
+
+    if execution_strategy is None:
+        raise ArgumentError(execution_strategy, 'Execution strategy cannot be None')
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:119]: In Cirq, the convention is to use `ValueError` instead of `ArgumentError` to signal that the value of an argument is wrong. Please change to `ValueError`.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 4]
```diff
@@ -12,10 +12,13 @@
...
 
+from argparse import ArgumentError
 from typing import DefaultDict, Dict, Sequence, TYPE_CHECKING, Optional
 
 import abc
 from collections import defaultdict
+import cirq
+
 
 from cirq import circuits, devices, ops, protocols
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:23]:
```suggestion

from cirq import circuits, devices, ops, protocols, transformers
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 5]
```diff
@@ -100,6 +106,47 @@ def optimization_at(
...
+    *,
+    execution_strategy: ExecutionStrategy = None,
+    context: Optional['cirq.TransformerContext'] = None,
+) -> None:
+    """Executes an acquaintance strategy"""
+
+    if execution_strategy is None:
+        raise ArgumentError(execution_strategy, 'Execution strategy cannot be None')
+
+    mapping = execution_strategy.initial_mapping.copy()
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:121]: Do we need this copy?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Looks like a good start, I've left some comments. Please let me know if you have questions!

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 5 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:121]: initial_mapping is just a LogicalMapping (a dictionary of integers). But it is mutated after every call to `op.gate.update_mapping(mapping, op.qubits)`. Since we're stateless I guess we can just pass around the same dictionary without creating copies like in StrategyExecutor.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 6]
```diff
@@ -64,6 +67,9 @@ def __call__(self, *args, **kwargs):
         return StrategyExecutor(self)(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:67]: I made this change. Now executor_test.py tests the new function instead of the old one but when I run it, I am getting the error message:
TypeError: Operation without a known matrix or decomposition: cirq.contrib.acquaintance.AcquaintanceOpportunityGate(num_qubits=5).on(cirq.LineQubit(0), cirq.LineQubit(1), cirq.LineQubit(2), cirq.LineQubit(3), cirq.LineQubit(4)).

I try to print the given circuit before and after the mapping and I get something like
0: ───█───
      │
1: ───█───
      │
2: ───█───
      │
3: ───█───
      │
4: ───█───

After mapping
0: ───█───0↦0───
      │   │
1: ───█───1↦1───
      │   │
2: ───█───2↦2───
      │   │
3: ───█───3↦3───
      │   │
4: ───█───4↦4───
(using num_qubits=5 and acquaintance_size=5)

From the error message, it is failing because it is unable to represent the circuit into a unitary. Is this because of the arrows ↦ ??  From what I understand the arrows ↦ signify a mapping from a qubit to another? But I am not sure either.

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 4 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:23]: But don't we need to import cirq for `@cirq.transformer` and `@cirq._compat.deprecated_class`? Or should I try to import the modules for these too?

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 7]
```diff
@@ -29,9 +30,6 @@
 )
 from cirq.contrib.acquaintance.mutation_utils import expose_acquaintance_gates
 
-if TYPE_CHECKING:
-    import cirq
-
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:34]: `import cirq` should only be done for type checking. So revert this change and remove the `import cirq` statement on top.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 8]
```diff
@@ -12,12 +12,13 @@
...
 
-from typing import DefaultDict, Dict, Sequence, TYPE_CHECKING, Optional
+from typing import DefaultDict, Dict, Sequence, Optional
 
 import abc
 from collections import defaultdict
 
-from cirq import circuits, devices, ops, protocols
+from cirq import circuits, devices, ops, protocols, transformers
+import cirq
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:21]:
```suggestion
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 9]
```diff
@@ -12,12 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import DefaultDict, Dict, Sequence, TYPE_CHECKING, Optional
+from typing import DefaultDict, Dict, Sequence, Optional
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:15]:
```suggestion
from typing import DefaultDict, Dict, Sequence, TYPE_CHECKING, Optional
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 10]
```diff
@@ -61,9 +59,26 @@ def get_operations(
         """Gets the logical operations to apply to qubits."""
 
     def __call__(self, *args, **kwargs):
-        return StrategyExecutor(self)(*args, **kwargs)
+        """Returns a copy of the modified circuit and the final mapping of
+        logical indices to qubits
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:63]: Update stale docstrings.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 11]
```diff
@@ -61,9 +59,26 @@ def get_operations(
...
+        """
+        if not isinstance(args[0], circuits.AbstractCircuit):
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:65]:
```suggestion
        if len(args) < 1 or not isinstance(args[0], circuits.AbstractCircuit):
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 12]
```diff
@@ -61,9 +59,26 @@ def get_operations(
...
+            raise ValueError(
+                (
+                    "To call ExecutionStrategy, an argument of type "
+                    "'cirq.Circuit' must be passed in as the first non-keyword argument"
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:69]:
```suggestion
                    "'cirq.AbstractCircuit' must be passed in as the first non-keyword argument"
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 13]
```diff
@@ -61,9 +59,26 @@ def get_operations(
...
+            )
+        input_circuit = args[0]
+        strategy = StrategyExecutorTransformer(self)
+        final_circuit = strategy(input_circuit, **kwargs)
+        input_circuit._moments = final_circuit._moments
+        return strategy.get_mapping()
 
 
+@cirq._compat.deprecated_class(
+    deadline='v1.0', fix='Use cirq.contrib.acquaintance.strategy_executor.'
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:80]:
```suggestion
    deadline='v1.0', fix='Use cirq.contrib.acquaintance.StrategyExecutorTransformer.'
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 14]
```diff
@@ -100,6 +115,65 @@ def optimization_at(
...
 
+class StrategyExecutorTransformer:
+    """Executes an acquaintance strategy."""
+
+    def __init__(self, execution_strategy: ExecutionStrategy) -> None:
+        self.execution_strategy = execution_strategy
+        self.mapping = execution_strategy.initial_mapping.copy()
+
+    def __call__(
+        self, circuit: 'cirq.Circuit', context: Optional['cirq.TransformerContext'] = None
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:126]:
```suggestion
        self, circuit: 'cirq.AbstractCircuit', context: Optional['cirq.TransformerContext'] = None
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 15]
```diff
@@ -100,6 +115,65 @@ def optimization_at(
...
+    ) -> circuits.AbstractCircuit:
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:127]:
```suggestion
    ) -> circuits.Circuit:
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 16]
```diff
@@ -100,6 +115,65 @@ def optimization_at(
...
+
+        Args:
+            circuit: Input circuit to transform.
+            context: `cirq.TransformerContext` storing common configurable
+              options for transformers.
+
+        Raises:
+            ValueError: if execution_strategy is None
+            TypeError: if execution_strategy neither an instance of
+              AcquaintanceOpportunityGate or PermutationGate
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:139]: These value errors are checking the type of `execution_strategy`; so they should be raised in the `__init__` method itself when instantiating the class with incorrect parameters.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 17]
```diff
@@ -100,6 +115,65 @@ def optimization_at(
...
+            raise ValueError('execution_strategy cannot be None')
+        expose_acquaintance_gates(circuit)
+        return transformers.map_operations(
+            circuit=circuit,
+            map_func=self.map_func,
+            deep=context.deep if context else False,
+            tags_to_ignore=context.tags_to_ignore if context else (),
+        )
+
+    def get_mapping(self):
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:156]: Add return type annotations.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 18]
```diff
@@ -100,6 +115,65 @@ def optimization_at(
         )
 
 
+class StrategyExecutorTransformer:
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:118]: Add tests for `StrategyExecutorTransformer`

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Left another round of comments.

**Comment by ammareltigani** [comment]
```I'm not sure why some of the tests are failing, here is my output locally when I run `./check/all`

TOTAL                                                                                                               95468   2458    97%

=========================== short test summary info ============================
FAILED dev_tools/docker_test.py::test_docker_pre - AssertionError: assert 1 == 0
FAILED dev_tools/docker_test.py::test_docker_stable - AssertionError: assert ...
2 failed, 16372 passed, 93 skipped, 132 xfailed, 504 warnings in 277.64s (0:04:37)
The annotate command will be removed in a future version.
Get in touch if you still use it: ned@nedbatchelder.com

All touched lines covered

Running doctest
Loading tests   

Executing tests 

Passed: 0```

**Comment by ammareltigani** [comment]
> I'm not sure why some of the tests are failing, here is my output locally when I run `./check/all`
> 
> ```
> TOTAL                                                                                                               95468   2458    97%
> 
> =========================== short test summary info ============================
> FAILED dev_tools/docker_test.py::test_docker_pre - AssertionError: assert 1 == 0
> FAILED dev_tools/docker_test.py::test_docker_stable - AssertionError: assert ...
> 2 failed, 16372 passed, 93 skipped, 132 xfailed, 504 warnings in 277.64s (0:04:37)
> The annotate command will be removed in a future version.
> Get in touch if you still use it: ned@nedbatchelder.com
> 
> All touched lines covered
> 
> Running doctest
> Loading tests   
> 
> Executing tests 
> 
> Passed: 0
> ```

Just realized now that these failures are not triggered by my code but are part of a larger issue with Qiskit.Aer dependency that is currently being deliberated by Palov and Cheng. I'll wait for their PR fixing that before merging this one.

**Comment by pavoljuhas** [comment]
> ...
> FAILED dev_tools/docker_test.py::test_docker_pre - AssertionError: assert 1 == 0
> FAILED dev_tools/docker_test.py::test_docker_stable - AssertionError: assert ...
> 2 failed, 16372 passed, 93 skipped, 132 xfailed, 504 warnings in 277.64s (0:04:37)

I also get those 2 failures when running tests locally, I suspect they need docker installation to pass.
You can ignore those two unless your PR makes some changes to the Dockerfile.

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 19]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+    """Executes an acquaintance strategy."""
+
+    def __init__(self, execution_strategy: ExecutionStrategy) -> None:
+        """Initializes transformer.
+
+        Raises:
+            ValueError: if execution_strategy is None.
+        """
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:128]: nit: Add `Args` section to init docstring. 
```suggestion
        """Initializes transformer.
        
        Args:
            execution_strategy: The `ExecutionStrategy` to execute. 

        Raises:
            ValueError: if execution_strategy is None.
        """
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 20]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+
+        if execution_strategy is None:
+            raise ValueError('execution_strategy cannot be None')
+        self.execution_strategy = execution_strategy
+        self.mapping = execution_strategy.initial_mapping.copy()
+
+    def __call__(
+        self, circuit: circuits.Circuit, context: Optional['cirq.TransformerContext'] = None
+    ) -> circuits.Circuit:
+        """Executes an acquaintance strategy using primitive map function and
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:138]:
```suggestion
        """Executes an acquaintance strategy using `cirq.map_operations` and
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 21]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+          mutates initial mapping.
+
+        Args:
+            circuit: 'circuits.Circuit' input circuit to transform.
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:142]:
```suggestion
            circuit: 'cirq.Circuit' input circuit to transform.
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 22]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+            ValueError: if execution_strategy is None.
+        """
+
+        if execution_strategy is None:
+            raise ValueError('execution_strategy cannot be None')
+        self.execution_strategy = execution_strategy
+        self.mapping = execution_strategy.initial_mapping.copy()
+
+    def __call__(
+        self, circuit: circuits.Circuit, context: Optional['cirq.TransformerContext'] = None
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:136]:
```suggestion
        self, circuit: circuits.AbstractCircuit, context: Optional['cirq.TransformerContext'] = None
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 23]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
         )
 
 
+class StrategyExecutorTransformer:
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:120]: Decorate the class with `transformers.transformer` decorator.
```suggestion
@transformers.transformer
class StrategyExecutorTransformer:
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 24]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+            circuit: 'circuits.Circuit' input circuit to transform.
+            context: `cirq.TransformerContext` storing common configurable
+              options for transformers.
+
+        Returns:
+            A copy of the modified circuit after executing an acquaintance
+              strategy on all instances of AcquaintanceOpportunityGate
+        """
+
+        expose_acquaintance_gates(circuit)
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:151]: For this method to be able to work on `cirq.AbstractCircuit`, you'd have to update the `expose_acquaintance_gates(circuit)` because this call currently expects a `cirq.Circuit` to work on. 

This can be done easily by:

```suggestion
        circuit = transformers.expand_composite(circuit, no_decomp= expose_acquaintance_gates.no_decomp)
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 25]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+            circuit=circuit,
+            map_func=self.map_func,
+            deep=context.deep if context else False,
+            tags_to_ignore=context.tags_to_ignore if context else (),
+        )
+
+    def get_mapping(self) -> LogicalMapping:
+        return self.mapping.copy()
+
+    def map_func(self, op: 'cirq.Operation', index) -> 'cirq.OP_TREE':
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:162]: nit: Change to private method. 

```suggestion
    def _map_func(self, op: 'cirq.Operation', index) -> 'cirq.OP_TREE':
```

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 26]
```diff
@@ -100,6 +117,67 @@ def optimization_at(
...
+        expose_acquaintance_gates(circuit)
+        return transformers.map_operations_and_unroll(
+            circuit=circuit,
+            map_func=self.map_func,
+            deep=context.deep if context else False,
+            tags_to_ignore=context.tags_to_ignore if context else (),
+        )
+
+    def get_mapping(self) -> LogicalMapping:
+        return self.mapping.copy()
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:160]: Do we need to return a copy here? 
Also, let's change this to a property `mapping` and rename `self.mapping` to `self._mapping`. 

```suggestion

    @property
    def mapping(self) -> LogicalMapping:
        return self._mapping
```

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Looking pretty good; a final round of comments!

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 21 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:142]: always reference the base module for an import in docstring?

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 26 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:160]: I don't think so.

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 24 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py:151]: Okay!

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 21 above)
Comment (tanujkhattar) [reply cirq-core/cirq/contrib/acquaintance/executor.py]: Yes, the docstrings should always be `cirq.<object_name>` because in this format, the docs infrastructure automatically generates hyperlinks to reference pages of the objects on the website.

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 26 above)
Comment (tanujkhattar) [reply cirq-core/cirq/contrib/acquaintance/executor.py]: Can you add a `@property` decorator as well to make it a property instead of method?

File: cirq-core/cirq/contrib/acquaintance/executor.py [hunk 27]
```diff
@@ -100,6 +117,74 @@ def optimization_at(
...
+        circuit = transformers.expand_composite(
+            circuit, no_decomp=expose_acquaintance_gates.no_decomp
+        )
+        expose_acquaintance_gates(circuit)
+        return transformers.map_operations_and_unroll(
+            circuit=circuit,
+            map_func=self._map_func,
+            deep=context.deep if context else False,
+            tags_to_ignore=context.tags_to_ignore if context else (),
+        )
```
Comment (tanujkhattar) [inline cirq-core/cirq/contrib/acquaintance/executor.py:164]: Add an `unfreeze` to convert the abstract circuit into a circuit and resolve the typing error. 
```suggestion
        ).unfreeze(copy=False)
```

File: cirq-core/cirq/contrib/acquaintance/executor.py (hunk 21 above)
Comment (ammareltigani) [reply cirq-core/cirq/contrib/acquaintance/executor.py]: Good to know. Thanks!

**Review by tanujkhattar — APPROVED** [review]
Looks great, thanks for all the iterations!
